app/controllers/recryptrequests_controller.rb

```diff
@@ -44,20 +71,26 @@ def create_recrypt_request
     redirect_to session_destroy_path
   end
 
-  def recrypt_private_key
-    if current_user.recrypt_private_key!(params[:new_password], params[:old_password])
+  def recrypt_private_key(new_password: nil)
+    new_password ||= params[:new_password]
+    if current_user.recrypt_private_key!(new_password, params[:old_password])
+      update_provider
       flash[:notice] = t('flashes.recryptrequests.recrypted')
-      return redirect_to session_destroy_path
+      return redirect_to sso_path
     end
 
     flash[:error] = current_user.errors.full_messages.join
-    redirect_to recryptrequests_new_ldap_password_path
+    redirect_to user_authenticator.recrypt_path
   end
 
   def user_authenticator
     Authentication::UserAuthenticator.init(
-      username: current_user.username,
+      username: current_user.try(:username),
```

Comment:
Do we even want to instantiate an authenticator if the username is nil?

Reply:
Changed it. Was from the first implementation